packages/backend/src/index.ts

```diff
@@ -47,36 +47,8 @@ function makeCreateEnv(loadedConfigs: AppConfig[]) {
 
   return (plugin: string): PluginEnvironment => {
     const logger = getRootLogger().child({ type: 'plugin', plugin });
-    // Supported DBs are sqlite and postgres
-    const isPg = [
-      'POSTGRES_USER',
-      'POSTGRES_HOST',
-      'POSTGRES_PASSWORD',
-    ].every(key => config.getOptional(`backend.${key}`));
-
-    let knexConfig;
-
-    if (isPg) {
-      knexConfig = {
-        client: 'pg',
-        useNullAsDefault: true,
-        connection: {
-          port: config.getOptionalNumber('backend.POSTGRES_PORT'),
-          host: config.getString('backend.POSTGRES_HOST'),
-          user: config.getString('backend.POSTGRES_USER'),
-          password: config.getString('backend.POSTGRES_PASSWORD'),
-          database: `backstage_plugin_${plugin}`,
-        } as PgConnectionConfig,
-      };
-    } else {
-      knexConfig = {
-        client: 'sqlite3',
-        connection: ':memory:',
-        useNullAsDefault: true,
-      };
-    }
-
-    const database = knex(knexConfig);
+    const dbConfig = config.getConfig('backend.database');
+    const database = createDatabase(dbConfig, `backstage_plugin_${plugin}`);
```

Comment:
I think the explicit database name is not necessarily what we want per default. We want to support different scenarios - cohosting on one db, separating by schema, separating by database instance - and I was hoping that that would eventually be possible in an entirely config-driven way.

Reply:
That would be nice. For our current instance, I think the schema based approach would be more ideal (at least in the near term). Figuring out how migrations would work for both schema vs separate databases seems a bit tricky. I was hoping to do a follow up on migrations, but feel that has many more nuances. However, maybe we can define the API here to support it down the road. Perhaps something like:

```yaml
# app-config.yaml
backend:
  database:
    client: pg
    schemaSettings: {} # optional with some defaults
    connection:
      # ...
```

Reply:
Hm, will migrations be affected by that? Haven't thought about it frankly.

Reply:
I think this idea is just an additive, so wouldn't need to block this code for being useful and extendable. But I did comment in #1882 about changing this to pass in an object that gets merged rather then just the db name.